src/rgw/rgw_sal_d3n.cc

```diff
@@ -0,0 +1,269 @@
+{
+  std::unique_ptr<Object> obj = next->get_object(k);
+  return std::make_unique<D3NFilterObject>(std::move(obj), this);
+}
+
+int D3NFilterDriver::get_bucket(const DoutPrefixProvider* dpp, User* u, const rgw_bucket& b, std::unique_ptr<Bucket>* bucket, optional_yield y)
+{
+  std::unique_ptr<Bucket> bp;
+  int ret = next->get_bucket(dpp, u, b, &bp, y);
+  if (ret < 0) {
+    return ret;
+  }
+
+  Bucket* d3n_bp = new D3NFilterBucket(this, std::move(bp), u);
+  bucket->reset(d3n_bp);
+  return 0;
+}
+
+int D3NFilterDriver::get_bucket(User* u, const RGWBucketInfo& i, std::unique_ptr<Bucket>* bucket)
+{
+  std::unique_ptr<Bucket> bp;
+  int ret = next->get_bucket(u, i, &bp);
+  if (ret < 0) {
+    return ret;
+  }
+
+  Bucket* d3n_bp = new D3NFilterBucket(this, std::move(bp), u);
+  bucket->reset(d3n_bp);
+  return 0;
+}
+
+int D3NFilterDriver::get_bucket(const DoutPrefixProvider* dpp, User* u, const std::string& tenant, const std::string& name, std::unique_ptr<Bucket>* bucket, optional_yield y)
+{
+  rgw_bucket b;
+
+  b.tenant = tenant;
+  b.name = name;
+
+  return get_bucket(dpp, u, b, bucket, y);
+}
+
+std::unique_ptr<Object> D3NFilterBucket::get_object(const rgw_obj_key& k)
+{
+  std::unique_ptr<Object> obj = next->get_object(k);
+  return std::make_unique<D3NFilterObject>(std::move(obj), this, this->filter);
+}
+
+std::unique_ptr<Object::ReadOp> D3NFilterObject::get_read_op()
+{
+  std::unique_ptr<Object::ReadOp> rop = next->get_read_op();
+  return std::make_unique<D3NFilterReadOp>(std::move(rop), this, filter);
+}
+
+int D3NFilterObject::D3NFilterReadOp::prepare(optional_yield y, const DoutPrefixProvider* dpp)
+{
+  return next->prepare(y, dpp);
+}
+
+void D3NFilterObject::D3NFilterReadOp::cancel() {
+  aio->drain();
+}
+
+int D3NFilterObject::D3NFilterReadOp::drain(const DoutPrefixProvider* dpp) {
+  auto c = aio->wait();
+  while (!c.empty()) {
+    int r = flush(dpp, std::move(c));
+    if (r < 0) {
+      cancel();
+      return r;
+    }
+    c = aio->wait();
+  }
+  return flush(dpp, std::move(c));
+}
+
+int D3NFilterObject::D3NFilterReadOp::flush(const DoutPrefixProvider* dpp, rgw::AioResultList&& results) {
+  int r = rgw::check_for_errors(results);
+  if (r < 0) {
+    return r;
+  }
+  std::list<bufferlist> bl_list;
+
+  auto cmp = [](const auto& lhs, const auto& rhs) { return lhs.id < rhs.id; };
+  results.sort(cmp); // merge() requires results to be sorted first
+  completed.merge(results, cmp); // merge results in sorted order
+
+  while (!completed.empty() && completed.front().id == offset) {
+    auto bl = std::move(completed.front().data);
+
+    bl_list.push_back(bl);
+    offset += bl.length();
+    int r = client_cb->handle_data(bl, 0, bl.length());
+    if (r < 0) {
+      return r;
+    }
+
+    const std::lock_guard l(d3n_get_data.d3n_lock);
+    auto oid = completed.front().obj.oid;
+    if (bl.length() <= g_conf()->rgw_get_obj_max_req_size) {
```

Comment:
@cbodley @mkogan1 : do we really need to write to cache here? These async ops would have completed only if the parts would have been found in the datacache. Or are we writing them again, in case they would have been evicted by lru policy till these ops complete?

Reply:
@pritha-srivastava you're right that we shouldn't be writing to the cache for these cache read completions

it's the data we read below from `next->iterate()` that we should be using the populate this cache. we'll probably need to wrap the `RGWGetDataCB* cb` that we pass down to `next->iterate()`, so we can intercept the data it returns

Reply:
Thats already done in D3NFilterGetCB::handle_data. I will remove the writing to cache from here.